Spring Boot 1.5 and 2 support

ruifigueira opened the pull request:
This PR is to add support for both Spring Boot 1.5 and 2 as discussed in #268

chkal commented:
Hi @ruifigueira!

First of all: Sorry for my delayed response. I'm busy as hell and Togglz is really just a spare time project for which I unfortunately don't find much time anymore. So sorry about that.

I just had a quick look at your branch. The work you've done is impressive. The overall structure looks great. I like the idea of splitting the Spring Boot support into multiple modules with a common shared dependency and a common parent pom. So :+1: for that.

I just wonder whether the module names are a good choice. Naming the two modules `togglz-spring-boot-starter` and `togglz-spring-boot-legacy-starter` has a few disadvantages:

* The term `legacy` is rather vague. Especially if Spring Boot release 3.x at some point in the future we may end up with another legacy module. So this will get very confusing.
* I would like to give existing users an easy upgrade path if they want to update Togglz. With the current module naming users will have to update the name of the Togglz integration module from `togglz-spring-boot-starter` to `togglz-spring-boot-legacy-starter` if they update Togglz without also updating Spring Boot. So this will make updates more difficult.

So, I think we should choose one of the following two options:
* Keep the name `togglz-spring-boot-starter` for the 1.x module and add a new module `togglz-spring-boot2-starter`. IMO this would be a good way. Especially because updating Togglz for existing users will be straight forward and if people update their app to Spring Boot 2.x, they will have to adjust many packages anyway. Also, we already applied this pattern with the `togglz-junit` and `togglz-junit5` modules. 
* A variant of the option above would be to rename the current module to `togglz-spring-boot1-starter` and introduce a new `togglz-spring-boot2-starter`. This way users have to choose which version to use, but the name is very explicit about which Spring Boot version is required. The downside is that updating Togglz will require users to change module names again.

What do you think? In both cases we should also use one root package per module, which is required if we ever start using Java 9 modules. 

Unfortunately I'm not very familiar with Spring Boot, so reviewing the inner workings is difficult for me. Maybe @marceloverdijk could help here? But I would also be fine with just merging a first version and then wait for user feedback. ;-)

Thanks a lot for your help with this.

ruifigueira commented:
@chkal , glad to know I'm in the right track 👍 

I just tried a different approach, and I think that this is the one 😄 

So, the idea is to just import Togglz for both Spring 1.x and Spring 2.x with our `starter` module. To ensure only the right `AutoConfiguration`s are loaded, Spring has a very useful `@ConditionalOnClass` which checks if a class we pass can be found in the classpath and only then it actually loads the `AutoConfiguration`. Spring boot uses this mechanism a lot.

In our case, we can pass classes that exist in one version but not in the other.

I just tried it,  and it works great. If you want to try it, checkout https://github.com/ruifigueira/togglz-sb2-demo, follow the instructions and try to change spring boot and java version, they work fine in both scenarios (note: Spring Boot 2 requires Java 8).

I also renamed modules using `boot1` and `boot2` because of the package renaming (see below).

The structure now is:

* `togglz-spring-boot-common-autoconfigure`
* `togglz-spring-boot1-autoconfigure`
* `togglz-spring-boot2-autoconfigure`
* `togglz-spring-boot-starter`

And users only need to import `togglz-spring-boot-starter` into their project, they don't need to configure a different module depending on the Spring Boot version they have.

> In both cases we should also use one root package per module, which is required if we ever start using Java 9 modules.

I'm aware of that, I just didn't put any thought on that because I was just trying to put this working. I've changed them to `org.togglz.spring.boot1.autoconfigure` and `org.togglz.spring.boot2.autoconfigure` and left `common-autoconfigurer` with `org.togglz.spring.boot.autoconfigure`

chkal commented:
Thanks a lot for updating the pull requests. 

I've a few questions:
  * The build is failing on Travis and also on my local machine. Looks like there is something wrong with the bundlor plugin. See this [log](https://travis-ci.org/togglz/togglz/jobs/367450613).
  * Maybe `togglz-spring-boot-starter` should have a packaging type of `pom`. Currently it is `jar`, but it doesn't contain any classes.
  * It somehow feels weird to have both `togglz-spring-boot[1|2]-autoconfigure` modules on the classpath if you declare a dependency on `togglz-spring-boot-starter`. Couldn't this lead to classloading issues? As I mentioned before I'm not so familiar with Spring Boot, so I may be wrong about this.

Christian

ruifigueira commented:
> Maybe togglz-spring-boot-starter should have a packaging type of pom. Currently it is jar, but it doesn't contain any classes.

That's the way starters are implemented in spring boot, they basically aggregate all needed jars to set up a new functionality. See for instance the Spring Data JPA starter module:
https://github.com/spring-projects/spring-boot/tree/master/spring-boot-project/spring-boot-starters/spring-boot-starter-data-jpa

> The build is failing on Travis and also on my local machine. Looks like there is something wrong with the bundlor plugin. See this log.

I'll look into that.

> It somehow feels weird to have both togglz-spring-boot[1|2]-autoconfigure modules on the classpath if you declare a dependency on togglz-spring-boot-starter. 

@chkal, that's not an issue, as long as we ensure our configuration classes have those `@ConditionalOnClass` annotations, because Spring Boot doesn't try to load them at all if it cannot satisfy those conditions. See https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/autoconfigure/condition/ConditionalOnClass.html#name-- . In my demo project you can check that: depending on the Spring Boot version you set up, it will load one or the other `TogglzEndpointAutoConfguration`.

ruifigueira commented:
@chkal, another problem in the travis build is that Spring Boot requires Java 1.8+, so our `boot2` module does not compile against Java 1.7.

Is is ok if I remove `openjdk7` from travis? All modules except `boot2` have Java 1.7 has target version, so they can run in Java 1.7.

chkal commented:
@ruifigueira Thanks! Yeah, I also thought about dropping JDK7 support. And I'm actually fine with doing so for the next release.

I'll come back to you regarding your other comments later today or tomorrow as soon as I find some time to check them in detail.

ruifigueira commented:
@chkal and @marceloverdijk , Spring Boot team is actually discussing the right approach for supporting both versions: https://github.com/spring-projects/spring-boot/issues/12733.

 I just posted there with the solution I presented here, let's see what they suggest there before moving forward.

chkal commented:
Awesome. Thanks for the update.

chkal commented:
@ruifigueira Thanks a lot for commenting on spring-projects/spring-boot#12733. I agree that a common naming scheme would be great, and the Spring Boot team is the best party to give advice on this.

I definitely want to see you PR getting merged. But I think we should sort these naming and structural issues out before starting a detailed review.

Thanks a lot for your help on this! :beers:

ryangardner commented:
Are there people who aren't able to upgrade to spring-boot 2.0 but are able to upgrade to a newer togglz version? 

I was just about to drop togglz into one of my newer projects that is based on boot 2.0 and figured I'd check if it would "just work" or not and came across this issue - looks like you've done quite a bit of work on it already... 

Since you _already_ have it working with both a 1.x and 2.x - even though it might not be the prettiest solution - maybe releasing the version as it you have it now and settings some kind of deadline for no longer supporting spring boot 1.5 would be reasonable?

ruifigueira commented:
@chkal, I also opened https://github.com/spring-projects/spring-boot/issues/12941 to discuss the approach of a single `starter` / two `autoconfigure` modules, but maybe I'm just complicating things too much, and maybe we should wait for a new major Togglz release to replace Spring Boot version.

ryangardner commented:
I selfishly wouldn't mind a new spring boot version to be supported with a minor release - as I'm now working on a project that is spring-boot 2.0 based and am at the point where togglz would be very useful... And I'm sure I'm not the only one in this situation... 

Perhaps live with the two-autoconfigure modules / one-starter-module option for the time being and then in the next major release drop support for older spring boot versions?

MALPI commented on pom.xml:
```diff
@@ -348,12 +356,36 @@
 
+      <dependency>
+        <groupId>org.togglz</groupId>
+        <artifactId>togglz-spring-actuator-boot-starter</artifactId>
```
Not sure what this artifact is about, but there is a difference between `actuator` and `starter`. Actuator are basically exposing metrics and management functionality whereas starters are about adding features.

ruifigueira replied:
Geez, you are right, I don't even know where that came from :D Definitely a mistake, there's no module with that name neither, I'll fix it.

MALPI commented on spring-boot/boot1-autoconfigure/pom.xml:
```diff
@@ -0,0 +1,98 @@
+  <description>Togglz - Spring Boot 1.5 Autoconfigure</description>
+
+  <properties>
+    <spring-boot.version>1.5.4.RELEASE</spring-boot.version>
```
could be increased to `1.5.11`

MALPI commented on spring-boot/common-autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzFeature.java:
```diff
@@ -0,0 +1,64 @@
+
+	@Override
+	public int compareTo(TogglzFeature o) {
+		return name.compareTo(o.getName());
```
formatting looks strange here

ruifigueira commented:
Thanks @MALPI, I just fixed your comments.

chkal commented:
Hey @ruifigueira. I just read [this comment](https://github.com/spring-projects/spring-boot/issues/12941#issuecomment-383715714) by @philwebb and was wondering whether the common module should depend on the Spring Boot 2.0 instead of the 1.5 API. Spring Boot 2.0 clearly is the future, so depending on a newer version could make more sense. Any thoughts?

MALPI commented:
If I can support you @ruifigueira just let me know.

ruifigueira commented:
@chkal, common module only depends on Spring Boot 1.5 for compilation, because in runtime it can either run with Spring Boot 1.5 or Spring Boot 2. As @philwebb says, there is the risk of using classes that changed between versions, and  so we have three options:

1. just leave it as-is, but ensure that we are only using common bits spring boot 1.5 and 2
2. compile it against both spring boot 1.5 and 2 (using two different maven profiles, for instance)
3. remove `common-autoconfigure` module as suggested by @philwebb, and copy its classes to both `togglz-spring-boot1-autoconfigure` and `togglz-spring-boot2-autoconfigure`

Let me know you thoughts about this.

chkal commented:
I was about a fourth option:

  4. Let `common-autoconfigure` compile against Spring Boot 2.0 instead of 1.5

Wouldn't this be a good alternative to option (1)? However, I'm not strong on this one. I would be fine with both (1) and (4).

ruifigueira commented:
@chkal, I was trying to change the compilation of  `*-commons-autoconfigure` to use Spring Boot 2 instead but noticed that this would compromise its usage in Spring Boot 1.5 projects using Java 1.7 because Spring Boot 2 requires Java 1.8 to compile, so that jar would not load under 1.7 JVMs.

chkal commented on pom.xml:
```diff
@@ -49,6 +49,8 @@
     <osgi.bundles.version>2.5.0</osgi.bundles.version>
 
     <spring.version>3.2.17.RELEASE</spring.version>
+    
```
Do we really need these properties? Isn't `project.parent.basedir` the basedir of the sonatype parent? Can't we just use `../foobar` to get the basedir in the modules?

ruifigueira replied:
For one level structure, we wouldn't need that property, but because we have a two level structure for spring boot modules, we need that property, otherwise we would have to reconfigure `org.eclipse.virgo.bundlor.maven` for the 2nd-level modules.

In other words, for 1st-level modules we would need to have:

```xml
<osgiProfilePath>../java6-server.profile</osgiProfilePath>
```
and for 2nd-level modules it would be:


```xml
<osgiProfilePath>../../java6-server.profile</osgiProfilePath>
```

chkal commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/test/TogglzTestExecutionListener.java:
```diff
@@ -1,3 +1,19 @@
+ * limitations under the License.
+ */
+
 package org.togglz.spring.boot.test;
```
Should this also go into the `autoconfigure` package? I basically like the idea of having one unique package per module. Especially because duplicated packages between different modules will break with Java 9 modules.

ruifigueira replied:
I'm not sure about this one...  Honestly I think this is not related with auto-configuration, and I think that we probably should move this into another module, for instance, `togglz-spring-core`.

chkal replied:
Ah, I see. This class actually doesn't use any Spring Boot classes, correct? In this case `togglz-spring-core` would be a better place.

ruifigueira commented:
@chkal , regarding modules and package names, I agree with you, I only have doubts about the `togglz-spring-boot-legacy-starter`, because the [recommended naming convention is to use a `spring-boot-starter`](https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-developing-auto-configuration.html#boot-features-custom-starter-naming).

The `togglz-legacy-spring-boot-starter` was suggested by @snicol at https://github.com/togglz/togglz/issues/268#issuecomment-388368382.

chkal commented:
I see. What about `togglz-spring-boot1-starter`. So we would have:

* `togglz-spring-boot1-starter`: Current Spring Boot Version (2.x)
* `togglz-spring-boot-starter`: Old Spring Boot version (1.x)

Is this better? I'm also a bit undetermined about the name.

ruifigueira commented:
I took a better look on the Spring Boot conventions for package names, and we were not following them. Normally, actuator things like endpoints go inside an `actuate` package and related `AutoConfiguration` / `ContextConfiguration` classes go into a `actuate.autoconfigure` package.

For `togglz-spring-boot-autoconfigure`, I kept the package name `org.togglz.spring.boot.autoconfigure` because that only contains generic `AutoConfiguration` classes.

So, I propose for our starter and "legacy" starter package name to be `org.togglz.spring.boot.(legacy.)actuate`.

I didn't include the `starter` in the package name because Spring Boot doesn't have that convention. Normally starter projects don't have any code, they only expose dependencies, but in our case, that would mean an additional complexity that we don't actually need.

> What about togglz-spring-boot1-starter.

I prefer either `togglz-legacy-spring-boot-starter` or `togglz-spring-boot-legacy-starter` (this one doesn't follow the `*-spring-boot-starter` convention, but I think it's acceptable), and would avoid the inclusion of the version in the module name, after what we discussed with @snicoll.

snicoll commented:
> What about togglz-spring-boot1-starter.

Generation name in identifiers isn't great IMO so I'd avoid that if you can. ` togglz-legacy-spring-boot-starter` looks good, though I came with "legacy" out of inspiration. Perhaps there is a better name to refer to the "previous generation".

chkal commented:
@ruifigueira Thanks a for applying the changes to the package names. That's awesome.

Regarding the module name for the Spring Boot 1.x support: I'm not really convinced that including the major version in the module name is a bad thing. Especially as it would allow to keep this module even for a longer period. But maybe this is a matter of taste. However, I don't want to stand in the way. We can go with `togglz-legacy-spring-boot-starter` if the community prefers this name. :+1:

MALPI commented on spring-boot/autoconfigure/pom.xml:
```diff
@@ -4,16 +4,16 @@
-  <description>Togglz - Spring Boot Starter</description>
+  <artifactId>togglz-spring-boot-autoconfigure</artifactId>
+  <name>Togglz - Spring Boot Autoconfiguration</name>
+  <description>Togglz - Spring Boot Autoconfiguration </description>
```
Could be something like, Dependency to enable Togglz Autoconfiguration in Spring Boot environment.

MALPI commented on spring-boot/autoconfigure/pom.xml:
```diff
@@ -4,16 +4,16 @@
 
   <properties>
-    <spring-boot.version>1.5.4.RELEASE</spring-boot.version>
+    <spring-boot.version>1.5.12.RELEASE</spring-boot.version>
```
`1.5.13` is latest

MALPI commented on spring-boot/autoconfigure/pom.xml:
```diff
@@ -31,6 +31,12 @@
+      <timezone>CET</timezone>
+    </developer>
     <developer>
       <id>marceloverdijk</id>
```
This is redundant now?

MALPI commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzAutoConfiguration.java:
```diff
@@ -0,0 +1,251 @@
+/*
```
General question, @chkal is the license header needed in every file?

MALPI commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzAutoConfiguration.java:
```diff
@@ -0,0 +1,251 @@
+
+    @Configuration
+    @ConditionalOnMissingBean(FeatureProvider.class)
+    protected static class FeatureProviderConfiguration {
```
Why don't you move it to a separate class?

ruifigueira replied:
That is a nested `@Configuration`, it's more clean this way, otherwise we would have to `@Import` it, and besides it inheriths the `@Conditional`s from `TogglzAutoConfiguration`

MALPI commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzAutoConfiguration.java:
```diff
@@ -0,0 +1,251 @@
+
+    @Configuration
+    @ConditionalOnMissingBean(FeatureManager.class)
+    protected static class FeatureManagerConfiguration {
```
Same here...

ruifigueira replied:
Same as above.

MALPI commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzAutoConfiguration.java:
```diff
@@ -0,0 +1,251 @@
+        public FeatureManager featureManager(FeatureProvider featureProvider, List<StateRepository> stateRepositories,
+                UserProvider userProvider, ActivationStrategyProvider activationStrategyProvider) {
+            StateRepository stateRepository = null;
+            if (stateRepositories.size() == 1) {
```
Can the list be  `null`?

ruifigueira replied:
No, that's a `@Bean`, so it's parameters are injected by Spring, using matching beans. In case of a collection:

> In case of a Collection or Map dependency type, the container autowires all beans matching the declared value type.

as stated in `@Autowired` javadocs. If no bean matches, that collection will be empty.

MALPI commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzAutoConfiguration.java:
```diff
@@ -0,0 +1,251 @@
+            if (featuresFile != null) {
+                Resource resource = this.resourceLoader.getResource(featuresFile);
+                Integer minCheckInterval = properties.getFeaturesFileMinCheckInterval();
+                if (minCheckInterval != null) {
```
Maybe check as well if `minCheckInterval > 0`?

MALPI commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzAutoConfiguration.java:
```diff
@@ -0,0 +1,251 @@
+        public StateRepository stateRepository() throws IOException {
+            StateRepository stateRepository;
+            String featuresFile = properties.getFeaturesFile();
+            if (featuresFile != null) {
```
In order to keep number of nested blocks lower you could go for:

```
if (featuresFile == null) { 
 Map<String, FeatureSpec> features = properties.getFeatures();
                stateRepository = new InMemoryStateRepository();
                for (String name : features.keySet()) {
                    stateRepository.setFeatureState(features.get(name).state(name));
                }
} else {
....
}
```

ruifigueira replied:
I didn't get that. We would still need to handle the `featuresFile != null` in the `else` block,

MALPI commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzAutoConfiguration.java:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    @Configuration
+    @ConditionalOnMissingClass("org.springframework.security.config.annotation.web.configuration.EnableWebSecurity")
```
Why did you go for a String here?

ruifigueira replied:
It was already that way, and `@ConditionalOnMissingClass` only supports a string value, not a class.

MALPI commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzAutoConfiguration.java:
```diff
@@ -0,0 +1,251 @@
+    @Conditional(TogglzConsoleBaseConfiguration.OnConsoleAndNotUseManagementPort.class)
+    protected static class TogglzConsoleConfiguration extends TogglzConsoleBaseConfiguration {
+
+        public TogglzConsoleConfiguration(@Autowired TogglzProperties properties) {
```
No need for `@Autowired` when using constructor injection, it's done implicitly.

ruifigueira replied:
That's true, I'll fix it.

MALPI commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzConsoleBaseConfiguration.java:
```diff
@@ -0,0 +1,86 @@
+        return "";
+    }
+
+    static class OnConsoleEnabled extends AllNestedConditions {
```
Is there missing code for the methods?

ruifigueira replied:
I don't understand, which method?

MALPI commented on spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzProperties.java:
```diff
@@ -380,6 +380,12 @@ public void setRegisterFeatureInterceptor(boolean registerFeatureInterceptor) {
+	 *
+	 * @see <a href="https://docs.spring.io/spring-boot/docs/current/reference/html/production-ready-endpoints.html">Spring Boot 2 chapter on Endpoints</a>
+	 */
+	@Deprecated
```
I wouldn't mark it as `@Deprecated` since the user had the choice to go for version 1 or 2 of Spring Boot. If he/she actively decided for 1, there is a reason and probably he/she is aware of that.

MALPI replied:
Or do I miss something?

ruifigueira replied:
I think you're right on this one, unless we also deprecate the Spring Boot 1 support, we should keep this the way it was. More thoughts on that: https://github.com/togglz/togglz/issues/268#issuecomment-388368382

MALPI commented on spring-boot/legacy-starter/src/test/java/org/togglz/spring/boot/legacy/actuate/BaseTest.java:
```diff
@@ -0,0 +1,119 @@
+import org.togglz.spring.boot.legacy.actuate.autoconfigure.TogglzManagementContextConfiguration;
+
+public class BaseTest {
+
```
Why don't you make use of `@SpringBootTest`? In general I'm no fan of having a base class for Tests which is inherited by others since you create some kind of dependency to all tests which might have unwanted side effects on change.

ruifigueira replied:
I understand your point of view, but each test requires a different configuration, and using `@SpringBootTest` would not be practical. In the Spring Boot 2, it's not hard to remove the `BaseTest` class because it doesn't need to maintain any state, but in Spring Boot 1, is not as simple, because it manages the lifecycle of the `ApplicationContext`.

Maybe we can refactor these tests in a future PR?

MALPI commented on spring-boot/starter/pom.xml:
```diff
@@ -0,0 +1,97 @@
+
+  <artifactId>togglz-spring-boot-starter</artifactId>
+  <name>Togglz - Spring Boot 2 Starter</name>
+  <description>Togglz - Spring Boot 2 Starter</description>
```
Starter Dependency for Togglz in Spring Boot environment.

ruifigueira replied:
Same as above.

MALPI commented on spring-boot/starter/pom.xml:
```diff
@@ -0,0 +1,97 @@
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+      <dependency>
```
Probably only needed for Tests?

ruifigueira replied:
That is the BOM dependency for Spring Boot: https://docs.spring.io/spring-boot/docs/current/reference/html/using-boot-build-systems.html#using-boot-maven-without-a-parent

MALPI commented on spring-boot/starter/src/main/java/org/togglz/spring/boot/actuate/TogglzEndpoint.java:
```diff
@@ -0,0 +1,56 @@
+    private final FeatureManager featureManager;
+
+    public TogglzEndpoint(FeatureManager featureManager) {
+        Assert.notNull(featureManager, "FeatureManager must not be null");
```
Shouldn't you throw an error here? Assertions are only activated in production if certain flags are set in the JVM AFAIK?

ruifigueira commented:
Talking with @chkal, we decided to merge this branch. Thanks a lot for all your suggestions and help!

This will make it easier for the community to contribute to improve Spring Boot 2 support.

snicoll commented:
Nice work. Thanks!
